src/audio_core/cubeb_input.cpp

```diff
@@ -0,0 +1,145 @@
+};
+
+CubebInput::CubebInput() : impl(std::make_unique<Impl>()) {
+    if (cubeb_init(&impl->ctx, "Citra Input", nullptr) != CUBEB_OK) {
+        LOG_ERROR(Audio, "cubeb_init failed! Mic will not work properly");
+        return;
+    }
+    impl->sample_queue = std::make_unique<Frontend::Mic::SampleQueue>();
+}
+
+CubebInput::~CubebInput() {
+    if (!impl->ctx)
+        return;
+
+    if (cubeb_stream_stop(impl->stream) != CUBEB_OK) {
+        LOG_ERROR(Audio, "Error stopping cubeb input stream.");
+    }
+
+    cubeb_destroy(impl->ctx);
+}
+
+void CubebInput::StartSampling(Frontend::Mic::Parameters params) {
+    // Cubeb apparently only supports signed 16 bit PCM (and float32 which the 3ds doesn't support)
+    // TODO resample the input stream
+    if (params.sign == Frontend::Mic::Signedness::Unsigned) {
+        LOG_ERROR(Audio,
+                  "Application requested unsupported unsigned pcm format. Falling back to signed");
+    }
+    if (params.sample_size != 16) {
+        LOG_ERROR(Audio,
+                  "Application requested unsupported 8 bit pcm format. Falling back to 16 bits");
+    }
+
+    parameters = params;
+    is_sampling = true;
+
+    impl->sample_size_in_bytes = 2;
+
+    cubeb_devid input_device = nullptr;
+    cubeb_stream_params input_params;
+    input_params.channels = 1;
+    input_params.layout = CUBEB_LAYOUT_UNDEFINED;
+    input_params.prefs = CUBEB_STREAM_PREF_NONE;
+    input_params.format = CUBEB_SAMPLE_S16LE;
+    input_params.rate = params.sample_rate;
+
+    u32 latency_frames;
+    if (cubeb_get_min_latency(impl->ctx, &input_params, &latency_frames) != CUBEB_OK) {
+        LOG_ERROR(Audio, "Could not get minimum latency");
+    }
+
+    if (cubeb_stream_init(impl->ctx, &impl->stream, "Citra Microphone", input_device, &input_params,
+                          nullptr, nullptr, latency_frames, Impl::DataCallback, Impl::StateCallback,
+                          impl.get()) != CUBEB_OK) {
+        LOG_CRITICAL(Audio, "Error creating cubeb input stream");
+    }
+
+    cubeb_stream_start(impl->stream);
+}
+
+void CubebInput::StopSampling() {
+    if (impl->stream) {
+        cubeb_stream_stop(impl->stream);
+    }
+    is_sampling = false;
+}
+
+void CubebInput::AdjustSampleRate(u32 sample_rate) {
+    // TODO This should restart the stream with the new sample rate
+    LOG_ERROR(Audio, "AdjustSampleRate unimplemented!");
+}
+
+Frontend::Mic::Samples CubebInput::Read() {
+    Frontend::Mic::Samples samples{};
+    Frontend::Mic::Samples queue;
+    while (impl->sample_queue->Pop(queue)) {
+        samples.insert(samples.end(), queue.begin(), queue.end());
+    }
+    return samples;
+}
+
+long CubebInput::Impl::DataCallback(cubeb_stream* stream, void* user_data, const void* input_buffer,
+                                    void* output_buffer, long num_frames) {
+    Impl* impl = static_cast<Impl*>(user_data);
+    u8 const* data = reinterpret_cast<u8 const*>(input_buffer);
```

Comment:
```suggestion
    const u8* data = reinterpret_cast<const u8*>(input_buffer);
```

This can also be moved after the conditional check below.